### PR TITLE
Add js env test kit and fix some bugs.

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Review by @sjrd

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala [hunk 1]
```diff
@@ -38,20 +44,29 @@ class SeleniumComJSRunner(browserProvider: SeleniumBrowser,
 
   def send(msg: String): Unit = {
     awaitForBrowser()
-    browser.getWebDriver.executeAsyncScript(sendScript, msg)
+    val encodedMsg = msg.replace("\u0000", "%00")
+    browser.getWebDriver.executeAsyncScript(sendScript, encodedMsg)
```
Comment (sjrd) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala:49]: What if `msg` contains the substring `"%00"`?

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumAsyncJSRunner.scala [hunk 2]
```diff
@@ -32,12 +32,8 @@ class SeleniumAsyncJSRunner(browserProvider: SeleniumBrowser,
   }
 
   override def stop(): Unit = synchronized {
-    if (!keepAlive || ignoreKeepAlive) {
-      future.onComplete { _ =>
-        browser.processConsoleLogs(console)
-        browser.close()
-      }
-    }
+    if (browser.isOpened && !keepAlive || ignoreKeepAlive)
```
Comment (sjrd) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumAsyncJSRunner.scala]: Add parentheses around the `&&` part. Don't rely on `&&` having higher precedence than `||`.

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumAsyncJSRunner.scala [hunk 3]
```diff
@@ -32,8 +32,8 @@ class SeleniumAsyncJSRunner(browserProvider: SeleniumBrowser,
   }
 
   override def stop(): Unit = synchronized {
-    if (!keepAlive || ignoreKeepAlive) {
-      future.onComplete { _ =>
+    if (browser.isOpened && !keepAlive || ignoreKeepAlive) {
```
Comment (sjrd) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumAsyncJSRunner.scala]: Add parentheses around the `&&` part. Don't rely on `&&` having higher precedence than `||`.

Comment (sjrd) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumAsyncJSRunner.scala]: Well actually I'm wondering whether it shouldn't be `browser.isOpened && (!keepAlive || ignoreKeepAlive)`.

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumAsyncJSRunner.scala [hunk 4]
```diff
@@ -32,8 +32,12 @@ class SeleniumAsyncJSRunner(browserProvider: SeleniumBrowser,
...
 
   override def stop(): Unit = synchronized {
-    if (browser.isOpened && !keepAlive || ignoreKeepAlive)
-      Try(browser.close())
+    if (browser.isOpened && !keepAlive || ignoreKeepAlive) {
+      Try {
+        browser.processConsoleLogs(console)
+        browser.close()
+      }
+    }
```
Comment (sjrd) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumAsyncJSRunner.scala]: Shouldn't this change be in the parent commit? (i.e., the Fix #53 one)

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala [hunk 5]
```diff
@@ -52,7 +60,12 @@ class SeleniumComJSRunner(browserProvider: SeleniumBrowser,
 
         case msg: String =>
           browser.processConsoleLogs(console)
-          msg.replace("%00", "\u0000")
+          if (msg.startsWith(MESSAGE_TAG))
+            msg.drop(MESSAGE_TAG.length).replace("%00", "\u0000")
```
Comment (sjrd) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala]: `substring` is more common than `drop` for `String`

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala [hunk 6]
```diff
@@ -38,20 +44,29 @@ class SeleniumComJSRunner(browserProvider: SeleniumBrowser,
...
+        throw new TimeoutException()
       browser.getWebDriver.executeAsyncScript(receiveScript) match {
         case null =>
           loop()
 
         case msg: String =>
           browser.processConsoleLogs(console)
-          msg
+          if (msg.startsWith(MESSAGE_TAG))
+            msg.drop(MESSAGE_TAG.length).replace("%00", "\u0000")
```
Comment (sjrd) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala]: `substring` is more common than `drop` for `String`

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumAsyncJSRunner.scala (hunk 3 above)
Comment (nicolasstucki) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumAsyncJSRunner.scala]: Actually it should be the second.

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumBrowser.scala [hunk 7]
```diff
@@ -48,4 +50,59 @@ trait SeleniumBrowser {
       """.stripMargin
     )
   )
+
+  /* Used to know if there exists and event that is going to run in the future
```
Comment (sjrd) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumBrowser.scala]: and -> an

**Comment by sjrd** [comment]
I am not convinced by the fix for #55. It would wait for timeouts alright, but it would still fail to wait for other events from the event loop. And in Selenium, there could be a lot such events: pending AJAX requests, waiting for `Promise` resolution, waiting for simulated actions on the DOM to trigger events, etc.

File: .travis.yml [hunk 8]
```diff
@@ -37,13 +40,13 @@ script:
 scala:
   - 2.10.6
   - 2.11.7
-  - 2.12.0-M3
+  - 2.12.0-M4
```
Comment (sjrd) [inline .travis.yml:43]: The version number bumps should be in their own commit.

File: build.sbt [hunk 9]
```diff
@@ -12,14 +12,14 @@ import com.typesafe.tools.mima.plugin.MimaKeys.{previousArtifact, binaryIssueFil
 val previousVersion = "0.1.1"
 
 val scalaVersionsUsedForPublishing: Set[String] =
-  Set("2.10.6", "2.11.7", "2.12.0-M3")
+  Set("2.10.6", "2.11.8", "2.12.0-M4")
```
Comment (sjrd) [inline build.sbt:15]: In this file too: the version number bumps should be in their own commit.

Comment (sjrd) [inline build.sbt:15]: Btw, that commit should also update the badge in the readme to 0.6.8+.

**Comment by sjrd** [comment]
That's all.

File: .travis.yml (hunk 8 above)
Comment (nicolasstucki) [inline .travis.yml:43]: Done

File: build.sbt (hunk 9 above)
Comment (nicolasstucki) [inline build.sbt:15]: Done

**Comment by nicolasstucki** [comment]
Removed the fix for #55 from this PR.

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala [hunk 10]
```diff
@@ -37,12 +41,18 @@ class SeleniumComJSRunner(browserProvider: SeleniumBrowser,
   }
 
   def send(msg: String): Unit = {
+    if (comClosed)
+      throw new ComJSEnv.ComClosedException
     awaitForBrowser()
-    browser.getWebDriver.executeAsyncScript(sendScript, msg)
+    val encodedMsg =
+      msg.replace("\u0007", "\u0007\u0007").replace('\u0000', '\u0007')
```
Comment (sjrd) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala]: What if `msg.contains("\u0000\u0000")`?

I suggest: `msg.replace("\u0007", "\u0007\u0007").replace("\u0000", "\u00070")`.

Btw, why `\u0007`? Why not something a bit more readable and conventional like `\\`: `msg.replace("\\", "\\\\").replace("\u0000", "\\0")`?

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala [hunk 11]
```diff
@@ -51,7 +61,18 @@ class SeleniumComJSRunner(browserProvider: SeleniumBrowser,
...
         case msg: String =>
           browser.processConsoleLogs(console)
-          msg
+          if (msg.startsWith(MESSAGE_TAG)) {
+            val taglessMsg = msg.substring(MESSAGE_TAG.length)
+            taglessMsg.replace("\u0007\u0007", "\u0000").map {
+              case '\u0000' => '\u0007'
+              case '\u0007' => '\u0000'
+              case c        => c
+            }
```
Comment (sjrd) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala]: The reverse transformation would be `taglessMsg.replace("\\0", "\u0000").replace("\\\\", "\\")`

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala [hunk 12]
```diff
@@ -93,10 +115,12 @@ class SeleniumComJSRunner(browserProvider: SeleniumBrowser,
          |      receiveBuf = null;
          |    },
          |    send: function(msg) {
-         |      sendMsgBufIn.push(msg);
+         |      var encodedMsg =
+         |        msg.split("\7").join("\7\7").split("\0").join("\7");
```
Comment (sjrd) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala]: There also exists `replace` in JS. The above code would give `msg.replace("\7", "\7\7").replace("\0", "\7")`, and my suggested alternative would be `msg.replace("\\", "\\\\").replace("\0", "\\0")`

**Comment by sjrd** [comment]
Typo in the commit message of "Encode '\0' while passing throug ..." -> "through"

**Comment by sjrd** [comment]
That's all.

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala (hunk 12 above)
Comment (nicolasstucki) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala]: The javascript replace only replaces the first occurrence of the string.

``` javascript
>  "abcabc".replace("a", "1")
<- "1bcabc"
```

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala (hunk 11 above)
Comment (nicolasstucki) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala]: If I start with `"\\0"`, it will encode it as `"\\\\0"`. Then this would decode it as `"\\\u0000"`.

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala (hunk 10 above)
Comment (nicolasstucki) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala]: I chose `"\u0007"` (bell character) because it is a character that nobody would ever use in javascript. This was chosen to minimize the overhead.

Comment (nicolasstucki) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala]: I believe that currently there is no issue with `msg.contains("\u0000\u0000")`, but only because of the order of the operations (quite fragile). I will use the `msg.replace("\u0007", "\u0007\u0007").replace("\u0000", "\u00070")` <del>which will also simplify a lot the decoding in JS</del>.

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala (hunk 12 above)
Comment (sjrd) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala]: Ah well ... I guess there's the regexp-based alternative, then. http://stackoverflow.com/a/17606289/1829647
Reportedly, it's faster.

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala (hunk 11 above)
Comment (sjrd) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala]: Arf, yes. That said, I don't quite understand how the current code is supposed to work. Maybe a good old loop/tailrec would be easier to get right.

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala (hunk 10 above)
Comment (sjrd) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala]: The bell character is also much more difficult to debug if I want to "see" the encoded messages. IMO the overhead argument is unfounded.

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala (hunk 12 above)
Comment (nicolasstucki) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala]: Just found that with the regex the is a flag to replace all matches.

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala (hunk 11 above)
Comment (nicolasstucki) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala]: The current on is quite fragile. I replaced the decoding with regex replace that does it in one pass, it should be clearer. 

I tried to avoid the the loops because i could not find a scheme that would work well in Scala and JS at the same time.

File: seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala (hunk 10 above)
Comment (nicolasstucki) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala]: True, just had this issue. I will replace it with `\`

Comment (nicolasstucki) [inline seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/SeleniumComJSRunner.scala]: Will have to use `&` instead because the `\` has too many layers of encoding and it becomes hard to understand.

**Comment by sjrd** [comment]
LGTM
